### PR TITLE
Ares merge #7: Add Ares-related game data and serialization behavior.

### PR DESCRIPTION
**Pull request by kberg**
This includes a change to how board spaces bonuses behave under Ares, though of course unused at the moment.

Depends on Ares merge 6 aka #1659

### CODE REVIEW

File: src/ares/AresHandler.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
+// Game.ts-specific behavior for Ares
```
Comment (chosta) [inline src/ares/AresHandler.ts:1]: I don't want to be a prick but do we really want this comment :)

File: src/ares/AresHandler.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+export class AresHandler {
+    public static hasHazardTile(space: ISpace): boolean {
+        return AresHandler.hazardSeverity(space) !== HazardSeverity.NONE;
+    }
+
+    public static hazardSeverity(space: ISpace): HazardSeverity {
+        const type = space.tile?.tileType;
+
+        switch(type) {
+        case TileType.DUST_STORM_MILD:
```
Comment (chosta) [inline src/ares/AresHandler.ts:21]: I'm the prick. Maybe use indentation here? The old argument "I've seen many projects and most of them use indentation on switch statements" :)

File: src/cards/ares/BioengineeringEnclosure.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  }
+
+  public canAct(player: Player): boolean {
+    // >1 because this player already has bioengineering enclosure.
+    return this.resourceCount > 0 &&
+      player.getResourceCards(this.resourceType).length > 1;
+  }
+
+  public action(player: Player, game: Game) {
+    // TODO(kberg): this code is a little brittle: if the player chooses not to select a
```
Comment (chosta) [inline src/cards/ares/BioengineeringEnclosure.ts:36]: Can't we force the player to select a card? 
I mean predators don't have an option for not munching another animal..

File: src/cards/ares/BioengineeringEnclosure.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    // card in the SelectResourceCard action, the player has still lost their animal.
+    this.resourceCount--;
+    game.log("${0} removed 1 animal from Bioengineering Enclosure.", b => b.player(player));
+
+    game.addInterrupt(new SelectResourceCard(player, game, this.resourceType, undefined, 1, undefined, [CardName.BIOENGINEERING_ENCLOSURE]));
+
+    return undefined;
+  }
+}
```
Comment (chosta) [inline src/cards/ares/BioengineeringEnclosure.ts:45]: Unrelated. Card design looks a bit off. Centerd content would be a nice improvement.

File: src/cards/ares/BiofertilizerFacility.ts [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    public play(player: Player, game: Game) {
+        player.addProduction(Resources.PLANTS, 1);
+
+        game.addResourceInterrupt(
+            player,
+            ResourceType.MICROBE,
+            2,
+            undefined);
+
+        return new SelectSpace(
```
Comment (chosta) [inline src/cards/ares/BiofertilizerFacility.ts:34]: No remarks. Just acknowledging how beautiful this part of the code is.

File: src/cards/ares/MetallicAsteroid.ts [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+            game.addTile(player, SpaceType.LAND, space, {
+              tileType: TileType.METALLIC_ASTEROID,
+              card: this.name
+            });
+            space.adjacency = {bonus: [SpaceBonus.STEEL]}
+            return undefined;
+          }
+      );
+    }
+}
```
Comment (chosta) [inline src/cards/ares/MetallicAsteroid.ts:37]: Unrelated. Card has double event tag (i had a bug in Card.ts and fixed it, maybe you've already seen it but just in case..) and tile placement looks a bit off.

File: src/cards/ares/SolarFarm.ts [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    public cost: number = 12;
+    public tags: Array<Tags> = [Tags.ENERGY, Tags.STEEL];
+    public cardType: CardType = CardType.AUTOMATED;
+    public name: CardName = CardName.SOLAR_FARM;
+    public play(player: Player, game: Game) {
+      return new SelectSpace(
+        "Select space for Solar Farm tile",
+        game.board.getAvailableSpacesOnLand(player),
+        (space: ISpace) => {
+            const plantsOnSpace = space.bonus.filter((b) => b === SpaceBonus.PLANT).length;
```
Comment (chosta) [inline src/cards/ares/SolarFarm.ts:24]: I'm not sure how you get "1 step for each plant tag on the area where you place the tile.". Meaning I don't understand the card :)

File: tests/cards/ares/BioengineeringEnclosure.spec.ts [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+  const scienceTagCard: IProjectCard = {
+    name: CardName.ACQUIRED_COMPANY,
+    cardType: CardType.ACTIVE,
+    cost: 0,
+    tags: [Tags.SCIENCE],
+    play: () => undefined
+  };
+
+  const animalHost: IProjectCard = {
+    name: CardName.ACQUIRED_SPACE_AGENCY,
```
Comment (chosta) [inline tests/cards/ares/BioengineeringEnclosure.spec.ts:26]: We are hosting animals on ACQUIRED_SPACE_AGENCY :)

File: tests/cards/ares/SolarFarm.spec.ts [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+        player = new Player("test", Color.BLUE, false);
+        game = new Game("foobar", [player, player], player, ARES_OPTIONS_WITH_HAZARDS);
+    });
+
+    it("Play", function () {
+        // Find the first spot to place a city.
+        const space = game.board.getAvailableSpacesForCity(player)[0];
+        // Hack the space to have a large number of plants, just to show a matching
+        // energy production bump - seven.
+        space.bonus = [
```
Comment (chosta) [inline tests/cards/ares/SolarFarm.spec.ts:28]: I should learn how this space.bonus works I guess..

**Review by chosta — APPROVED** [review]
Looking good. Minor issues with design (i'm looking at cards at $beta url, might be outdated).

File: src/ares/AresHandler.ts (hunk 2 above)
Comment (kberg) [reply src/ares/AresHandler.ts:21]: Huh! For some reason I thought we weren't. Yes. And you're fine.

File: src/ares/AresHandler.ts (hunk 1 above)
Comment (kberg) [reply src/ares/AresHandler.ts:1]: Removed.

File: src/cards/ares/BioengineeringEnclosure.ts (hunk 3 above)
Comment (kberg) [reply src/cards/ares/BioengineeringEnclosure.ts:36]: /shrug That was something I came up with when I got started. It could be a non-issue, particularly after @Lynesth's most recent change.

File: src/cards/ares/BioengineeringEnclosure.ts (hunk 4 above)
Comment (kberg) [reply src/cards/ares/BioengineeringEnclosure.ts:45]: Oh it most certainly would be. All the imagery needs a refresh. Most.

File: src/cards/ares/BiofertilizerFacility.ts (hunk 5 above)
Comment (kberg) [reply src/cards/ares/BiofertilizerFacility.ts:34]: Thanks!

File: src/cards/ares/MetallicAsteroid.ts (hunk 6 above)
Comment (kberg) [reply src/cards/ares/MetallicAsteroid.ts:37]: Most certainly the placement is off -- the visual elements are going to need a good scrubbing.

File: src/cards/ares/SolarFarm.ts (hunk 7 above)
Comment (kberg) [reply src/cards/ares/SolarFarm.ts:24]: Count the number of plants on the space. Then gain that much energy production. (Or energy units, I don't remember.)

Resolving because I think that's sufficient.

Comment (kberg) [reply src/cards/ares/SolarFarm.ts:24]: (lmk if I'm wrong.)

File: tests/cards/ares/BioengineeringEnclosure.spec.ts (hunk 8 above)
Comment (kberg) [reply tests/cards/ares/BioengineeringEnclosure.spec.ts:26]: I kind of like tests that don't rely on expected innards of other cards, so I intentionally pick arbitrarily.

Besides, I'm sure the agency could use a cat.

File: src/Game.ts [hunk 10]
```diff
@@ -175,6 +177,8 @@ export class Game implements ILoadable<SerializedGame, Game> {
           turmoilExtension: false,
           promoCardsOption: false,
           communityCardsOption: false,
+          aresExtension: false,
+          aresHazards: true,
```
Comment (bafolts) [inline src/Game.ts:181]: What does it represent when `aresExtension` is `false` but `aresHazards` is `true`?

Comment (kberg) [reply src/Game.ts:181]: I knew this was going to come up and I don't know why I didn't do anything about it.

`aresHazards` is meaningless without `aresExtension`, and `aresHazards` is not toggled on the main page, but only used to make certain tests easier to write by setting it `false`. Any thoughts

Comment (bafolts) [reply src/Game.ts:181]: Which tests did this make easier to write? As far as I can tell nothing in `/src/**/*.ts` is using the `aresHazard` property, we only set it and never read it. The `SolarFarm` unit tests seem to pass whether I use `ARES_OPTIONS_WITH_HAZARDS` or `ARES_OPTIONS_NO_HAZARDS`.

File: tests/ares/AresStorageTest.spec.ts [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+        const otherPlayer = new Player("other", Color.RED, false);
+        const game = new Game("foobar", [player, otherPlayer], player, ARES_OPTIONS_NO_HAZARDS);
+        const space = game.board.getSpaces(SpaceType.OCEAN).find(space => space.bonus.length > 0)!;
+        const spaceId = space.id;
+
+        expect(space.bonus.length).is.greaterThan(0);
+        game.addOceanTile(player, spaceId);
+        expect(space.bonus.length).is.eq(0);
+
+        const json = JSON.stringify(game, game.replacer);
```
Comment (bafolts) [inline tests/ares/AresStorageTest.spec.ts:22]: We should put this `JSON.stringify(game, game.replacer)` to a `toJSON()` function so we don't have to copy this logic and have `replacer` be public.

**Review by bafolts — CHANGES_REQUESTED** [review]
I believe this is the first time we have serialized a `Game` outside of the class. I think we should make a `toJSON` as the opposite of `loadFromJSON` to hide the implementation details of `replacer`.

File: src/Game.ts (hunk 10 above)
Comment (kberg) [reply src/Game.ts:181]: It's used in a lot of places: https://github.com/bafolts/terraforming-mars/search?q=ARES_OPTIONS_NO_HAZARDS&type=Code

This isn't all the hazards code, and the code currently under test was older code that didn't have to worry about hazards. Solar Farm is a good example of this.

File: tests/ares/AresStorageTest.spec.ts (hunk 11 above)
Comment (kberg) [reply tests/ares/AresStorageTest.spec.ts:22]: ack. I'm happy to do that -- you mind if it's a follow-up PR later?

File: src/Game.ts [hunk 12]
```diff
@@ -1480,6 +1484,15 @@ export class Game implements ILoadable<SerializedGame, Game> {
       if (tile.tileType === TileType.OCEAN) {
         space.player = undefined;
       }
+
+      // Must occur after all other onTilePlaced operations.
+      if (this.gameOptions.aresExtension) {
+        // Erasing the bonus means that overplacing doesn't regrant the bonus, or, for instance, make hazard spaces
+        // available for Mining Area or Solar Farm.
+        space.bonus = [];
```
Comment (Lynesth) [inline src/Game.ts:1492]: Won't this create an issue with Dry Deserts removing an Ocean tile ?
I should be able to get that bonus again when putting an Ocean at the same spot.

File: src/Game.ts [hunk 13]
```diff
@@ -1775,12 +1788,22 @@ export class Game implements ILoadable<SerializedGame, Game> {
...
+        space.adjacency = element.adjacency;
       });
 
+      if (this.gameOptions.aresExtension) {
+        this.aresData = d.aresData;
+        // Spaces with tiles no longer have the underlying bonuses. If there's overplacement, erasing them
+        // won't regrant them, nor will they be avialable for options like Mining Area.
+        this.board.spaces.forEach(space => {
+          if (space.tile !== undefined) {
+            space.bonus = [];
```
Comment (Lynesth) [inline src/Game.ts:1803]: Same issue as above.

File: tests/ares/AresStorageTest.spec.ts [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+import { Game } from "../../src/Game";
+import { Player } from "../../src/Player";
+import { ARES_OPTIONS_NO_HAZARDS } from "./AresTestHelper";
+import { SpaceType } from "../../src/SpaceType";
+import { expect } from "chai";
+import { SerializedGame } from "../../src/SerializedGame";
+import { TileType } from "../../src/TileType";
+
+describe("AresStorageTest", function () {
+    it("When reloading the board, covered spaces don't have bonuses", function() {
```
Comment (Lynesth) [inline tests/ares/AresStorageTest.spec.ts:11]: Same questioning as above regarding space bonuses.

File: src/Game.ts (hunk 12 above)
Comment (kberg) [reply src/Game.ts:1492]: Coming soon: https://github.com/kberg/terraforming-mars/blob/ares/tests/ares/OtherAresTests.spec.ts

Or eventually.

File: src/Game.ts (hunk 13 above)
Comment (kberg) [reply src/Game.ts:1803]: OK to resolve?

File: tests/ares/AresStorageTest.spec.ts (hunk 14 above)
Comment (kberg) [reply tests/ares/AresStorageTest.spec.ts:11]: OK to resolve?

Comment (kberg) [reply tests/ares/AresStorageTest.spec.ts:11]: Did I say I was going to sleep? I lied. :D

OK I'm going to sleep. Thanks for the review?

File: src/Game.ts (hunk 12 above)
Comment (kberg) [reply src/Game.ts:1492]: Oh no, that's not Dry Deserts. Good catch. Grr, hrm...

Comment (kberg) [reply src/Game.ts:1492]: I only considered Dry Deserts as it applies to _upgraded oceans._

I'll have to fix this,  yes. Since aresExpansion can't be enabled on a running server, I hope this can still be approved and merged, with a commitment for an upcoming PR to fix it (in the next day or tw.)

I'm particularly focused on trying to get momentum on the changes.

Comment (Lynesth) [reply src/Game.ts:1492]: Yeah I wouldn't mind. And even if that's not fixed before Ares makes it in, it really is an edge case non game breaking bug.

File: src/Game.ts (hunk 13 above)
Comment (Lynesth) [reply src/Game.ts:1803]: Yep

File: tests/ares/AresStorageTest.spec.ts (hunk 14 above)
Comment (Lynesth) [reply tests/ares/AresStorageTest.spec.ts:11]: You're welcome, resolve it :)

File: src/Game.ts (hunk 10 above)
Comment (bafolts) [reply src/Game.ts:181]: As far as I can tell the `aresHazards` option is not used.

https://github.com/bafolts/terraforming-mars/search?q=aresHazards&type=Code

The `ARES_OPTIONS_NO_HAZARDS` is being set but I can't see how we ever leverage `aresHazards` to take different code paths.

File: tests/ares/AresStorageTest.spec.ts (hunk 11 above)
Comment (bafolts) [reply tests/ares/AresStorageTest.spec.ts:22]: OK

**Review by bafolts — APPROVED** [review]

File: tests/ares/AresStorageTest.spec.ts (hunk 11 above)
Comment (kberg) [reply tests/ares/AresStorageTest.spec.ts:22]: Added a tracking issue.

File: src/Game.ts (hunk 10 above)
Comment (kberg) [reply src/Game.ts:181]: Good catch -- it is introduced here because I wanted to keep it with `aresExtension`  but is used in am10. I hadn't considered the disconnect, either. https://github.com/bafolts/terraforming-mars/pull/1691
